cpp/src/arrow/engine/substrait/serde_test.cc

```diff
@@ -3187,5 +3198,164 @@ TEST(Substrait, IsthmusPlan) {
+                  },
+                  "baseSchema": {
+                    "names": ["time", "key", "value1"],
+                    "struct": {
+                      "types": [
+                        {
+                          "i32": {
+                            "typeVariationReference": 0,
+                            "nullability": "NULLABILITY_NULLABLE"
+                          }
+                        },
+                        {
+                          "i32": {
+                            "typeVariationReference": 0,
+                            "nullability": "NULLABILITY_NULLABLE"
+                          }
+                        },
+                        {
+                          "fp64": {
+                            "typeVariationReference": 0,
+                            "nullability": "NULLABILITY_NULLABLE"
+                          }
+                        }
+                      ],
+                      "typeVariationReference": 0,
+                      "nullability": "NULLABILITY_REQUIRED"
+                    }
+                  },
+                  "namedTable": {
+                    "names": ["T1"]
+                  }
+                }
+              },
+              {
+                "read": {
+                  "common": {
+                    "direct": {
+                    }
+                  },
+                  "baseSchema": {
+                    "names": ["time", "key", "value2"],
+                    "struct": {
+                      "types": [
+                        {
+                          "i32": {
+                            "typeVariationReference": 0,
+                            "nullability": "NULLABILITY_NULLABLE"
+                          }
+                        },
+                        {
+                          "i32": {
+                            "typeVariationReference": 0,
+                            "nullability": "NULLABILITY_NULLABLE"
+                          }
+                        },
+                        {
+                          "fp64": {
+                            "typeVariationReference": 0,
+                            "nullability": "NULLABILITY_NULLABLE"
+                          }
+                        }
+                      ],
+                      "typeVariationReference": 0,
+                      "nullability": "NULLABILITY_REQUIRED"
+                    }
+                  },
+                  "namedTable": {
+                    "names": ["T2"]
+                  }
+                }
+              }
+            ],
+            "detail": {
+              "@type": "/arrow.substrait.AsOfJoinRel",
+              "on": {
+                "selection": {
+                  "directReference": {
+                    "structField": {
+                      "field": 0,
+                    }
+                  },
+                  "rootReference": {}
+                }
+              },
+              "by": [
```

Comment:
As noted in [this post](https://github.com/apache/arrow/pull/14385#discussion_r994865351), `AsOfJoinNode` code currently does not support distinct per-input references.

Reply:
I am confused - 
https://github.com/apache/arrow/blob/ec579df631deaa8f6186208ed2a4ebec00581dfa/cpp/src/arrow/compute/exec/options.h#L411

Looks like "by" is one per table and "on" is only one for all the tables. So shouldn't substrait message contains multiple entries for "by" key?

Reply:
No, both are common to all tables. The "by" field is a vector corresponding to columns, i.e., one can select multiple columns using this vector into the "by" key. If we want per-input field-references, we'd need a vector "on" and a vector-of-vector "by".

Reply:
I see - this makes sense now.

Unfortunately the limitation is too strict I think (requiring all input tables to have the same column index for the by key) 
The restriction for the on-key is OK - we have restrictions for "time" to be the first column. However, we don't have restriction on the "by-key" column - it can have different column index for different input tables.

I think we need to fix this problem in order to Asof Join to be useful - Since we are deciding the shape of the substrait message now - I think it makes sense to fix both message format and asof join options format in this PR.